Introduce `evicted-at`/`last-evicted` timestamps

evanlinjin opened the pull request:
Partially Fixes #1740.
Replaces #1765.
Replaces #1811.

### Description

This PR allows the receiving structures (`bdk_chain`, `bdk_wallet`) to detect and evict incoming transactions that are double spent (cancelled).

We add a new field to `TxUpdate` (`TxUpdate::evicted_ats`), which in turn, updates the `last_evicted` timestamps that are tracked/persisted by `TxGraph`. This is similar to how `TxUpdate::seen_ats` update the `last_seen` timestamp in `TxGraph`. Transactions with a `last_evicted` timestamp higher than their `last_seen` timestamp are considered evicted.

`SpkWithExpectedTxids` is introduced in `SpkClient` to track expected `Txid`s for each `spk`. During a sync, if any `Txid`s from `SpkWithExpectedTxids` are not in the current history of an `spk` obtained from the chain source, those `Txid`s are considered missing. Support for `SpkWithExpectedTxids` has been added to both `bdk_electrum` and `bdk_esplora` chain source crates.

The canonicalization algorithm is updated to disregard transactions with a `last_evicted` timestamp greater than or equal to their `last_seen` timestamp, except in cases where transitivity rules apply.

### Notes to the reviewers

This PR does not fix #1740 for block-by-block chain source (such as `bdk_bitcoind_rpc`). This work is done in a separate PR (#1857).

### Changelog notice

* Add `TxUpdate::evicted_ats` which tracks transactions that have been replaced and are no longer present in mempool.
* Change `TxGraph` to track `last_evicted` timestamps. This is when a transaction is last marked as missing from the mempool.
* The canonicalization algorithm now disregards transactions with a `last_evicted` timestamp greater than or equal to it's `last_seen` timestamp, except when a canonical descendant exists due to rules of transitivity.
* Add `SpkWithExpectedTxids` in `spk_client` which keeps track of expected `Txid`s for each `spk`.
* Change `bdk_electrum` and `bdk_esplora` to understand `SpkWithExpectedTxids`.
* Add `SyncRequestBuilder::expected_txids_of_spk` method which adds an association between `txid`s and `spk`s.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

evanlinjin committed:
feat(core): Add `TxUpdate::evicted_ats`

This is a set of evicted txs from the mempool.

evanlinjin committed:
feat(chain)!: Change `TxGraph` to understand evicted-at & last-evicted

The evicted-at and last-evicted timestamp informs the `TxGraph` when the
transaction was last deemed as missing (evicted) from the mempool.

The canonicalization algorithm is changed to disregard transactions with
a last-evicted timestamp greater or equal to it's last-seen timestamp.
The exception is when we have a canonical descendant due to rules of
transitivity.

Update rusqlite_impl to persist `last_evicted`.

Also update docs:
* Remove duplicate paragraphs about `ChangeSet`s.
* Add "Canonicalization" section which expands on methods that require
  canonicalization and the associated data types used in the
  canonicalization algorithm.

LagginTimes committed:
feat(core): Add expected txids to `SyncRequest` spks

The spk history returned from Electrum should have these txs present.
Any missing tx will be considered evicted from the mempool.

evanlinjin committed:
feat(chain): Add `TxGraph` methods that handle expected spk txids

* `TxGraph::try_list_expected_spk_txids`
* `TxGraph::list_expected_spk_txids`
* `IndexedTxGraph::try_list_expected_spk_txids`
* `IndexedTxGraph::list_expected_spk_txids`

Co-authored-by: valued mammal <valuedmammal@protonmail.com>
Co-authored-by: Wei Chen <wzc110@gmail.com>

evanlinjin committed:
feat(electrum): Handle spks with expected txids

Co-authored-by: Wei Chen <wzc110@gmail.com>

evanlinjin committed:
feat(esplora): Handle spks with expected txids

Also add `detect_receive_tx_cancel` test.
Also rm `miniscript` dependency.
Update ci to rm `miniscript/no-std` for `bdk_esplora`.

Co-authored-by: Wei Chen <wzc110@gmail.com>

nymius commented on crates/chain/src/tx_graph.rs:
```diff
@@ -16,23 +16,52 @@
-//! The [`ChangeSet`] reports changes made to a [`TxGraph`]; it can be used to either save to
-//! persistent storage, or to be applied to another [`TxGraph`].
+//! Conflicting transactions are allowed to coexist within a [`TxGraph`]. A process called
+//! canonicalization is required to get a conflict-free history of transactions.
```
I would emphasize view instead of history to avoid anyone thinking the canonicalization process is something definitive and irreversible. 
```suggestion
//! canonicalization is required to get a conflict-free view of transactions.
```

nymius commented on crates/chain/src/tx_graph.rs:
```diff
@@ -937,9 +1006,14 @@ impl<A: Anchor> TxGraph<A> {
+            .iter()
+            .copied()
+            .rev()
+            .filter(|(last_seen, txid)| !matches!(self.last_evicted.get(txid), Some(last_evicted) if last_evicted >= last_seen))
```
Only if you consider this is better than `!matches!(...)`
```suggestion
           .filter(|(last_seen, txid)| {
               self.last_evicted
                   .get(txid)
                   // Has the current tx been seen after its last eviction?
                   .map_or(true, |last_evicted| last_evicted < last_seen)
           })
```

nymius commented on crates/core/src/spk_client.rs:
```diff
@@ -153,6 +175,20 @@ impl<I> SyncRequestBuilder<I> {
         self
     }
 
+    /// Add transactions that are expected to exist under then given spks.
```
nit (typo I think)
```suggestion
    /// Add transactions that are expected to exist under the given spks.
```

nymius commented on crates/esplora/tests/async_ext.rs:
```diff
@@ -1,15 +1,139 @@
 
+// Ensure that a wallet can detect a malicious replacement of an incoming transaction.
+//
+// This checks that both the Electrum chain source and the receiving structures properly track the
```
nit
```suggestion
// This checks that both the Esplora chain source and the receiving structures properly track the
```

nymius commented on crates/esplora/tests/blocking_ext.rs:
```diff
@@ -1,15 +1,140 @@
 
+// Ensure that a wallet can detect a malicious replacement of an incoming transaction.
+//
+// This checks that both the Electrum chain source and the receiving structures properly track the
```
nit
```suggestion
// This checks that both the Esplora chain source and the receiving structures properly track the
```

nymius reviewed (commented):
cACK 9d6c795c

This unlocks functionality on demand and I think the approach is OK. Tests are thorough enough.

LagginTimes commented on crates/core/src/spk_client.rs:
```diff
@@ -86,6 +86,28 @@ impl SyncProgress {
+    /// [`Txid`]s that we expect to appear in the chain source's spk history response.
+    ///
+    /// Any transaction listed here that is missing form the spk history response shoud be
+    /// considred evicted from the mempool.
```
```suggestion
    /// Any transaction listed here that is missing from the spk history response should be
    /// considered evicted from the mempool.
```

LagginTimes commented on crates/chain/src/tx_graph.rs:
```diff
@@ -16,23 +16,52 @@
+//! * `last_seen` - This is the timestamp of when a transaction is last-seen in the mempool. This
+//!     value is updated by [`insert_seen_at`](TxGraph::insert_seen_at) and
+//!     [`apply_update`](TxGraph::apply_update). Transactions that are seen later have higher
+//!     priority than those that are seen earlier. `last_seen` values are transitive. Meaning that
```
```suggestion
//!     priority than those that are seen earlier. `last_seen` values are transitive, meaning that
```

LagginTimes commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1078,6 +1152,67 @@ impl<A: Anchor> TxGraph<A> {
+    ///
+    /// This is used to fill [`SyncRequestBuilder::expected_spk_txids`](bdk_core::spk_client::SyncRequestBuilder::expected_spk_txids).
+    ///
+    /// The spk index range can be contrained with `range`.
```
```suggestion
    /// The spk index range can be constrained with `range`.
```

LagginTimes reviewed (commented):
Nit: Minor typos and grammar fixes.

notmandatory commented on .github/workflows/cont_integration.yml:
```diff
@@ -94,7 +94,7 @@ jobs:
         working-directory: ./crates/esplora
         # TODO "--target thumbv6m-none-eabi" should work but currently does not
-        run: cargo check --no-default-features --features miniscript/no-std,bdk_chain/hashbrown
+        run: cargo check --no-default-features --features bdk_chain/hashbrown
```
👍 thanks for cleaning this up!

oleonardolima replied:
It raised my attention, shouldn't we be using `bdk_core/hashbrown` instead ?

notmandatory commented on crates/chain/src/tx_graph.rs:
```diff
@@ -16,23 +16,52 @@
-//! methods only consider transactions that are "canonical" (i.e., in the best chain or in mempool).
-//! We decide which transactions are canonical based on the transaction's anchors and the
-//! `last_seen` (as unconfirmed) timestamp.
+//! # Canonicalization
```
👍 Good to see these new docs as this is core to how the the `tx_graph` logic works.

notmandatory commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1107,6 +1245,9 @@ pub struct ChangeSet<A = ()> {
     /// Added last-seen unix timestamps of transactions.
     pub last_seen: BTreeMap<Txid, u64>,
+    /// Added timestamps of when a transaction is last evicted from the mempool.
+    #[cfg_attr(feature = "serde", serde(default))]
```
Is this `cfg_attr` only needed to load data from a pre-v2 serialized file?

evanlinjin replied:
Yes, this ensures backwards compatibility of our serialized changeset (as we promised).

notmandatory reviewed (approved):
ut ACK 1eb0c322ead9f454093abdfe08bf868f841d4a24

Thanks for the fix and all the new docs, looks like a solid solution. Before (or after) this goes in it would be nice to have some confirmation from @ErikDeSmedt that it fixes #1740 for him (at least using electrum and esplora).

oleonardolima commented on crates/core/src/tx_update.rs:
```diff
@@ -44,6 +44,12 @@ pub struct TxUpdate<A = ()> {
+    /// When transactions were discovered to be missing (evicted) from the mempool.
+    ///
+    /// [`SyncRequest::start_time`](crate::spk_client::SyncRequest::start_time) can be used to
+    /// provide the `seen_at` value.
```
nit:
``` suggestion
    /// provide the `evicted_at` value.
```

evanlinjin replied:
It can be used to provide both actually. Will change.

oleonardolima commented on crates/esplora/tests/common/mod.rs:
```diff
@@ -0,0 +1,14 @@
+use bdk_core::bitcoin::key::{Secp256k1, UntweakedPublicKey};
+use bdk_core::bitcoin::ScriptBuf;
+
+const PK_BYTES: &[u8] = &[
+    12, 244, 72, 4, 163, 4, 211, 81, 159, 82, 153, 123, 125, 74, 142, 40, 55, 237, 191, 231, 31,
+    114, 89, 165, 83, 141, 8, 203, 93, 240, 53, 101,
+];
+
+#[allow(dead_code)]
+pub fn get_test_spk() -> ScriptBuf {
+    let secp = Secp256k1::new();
+    let pk = UntweakedPublicKey::from_slice(PK_BYTES).expect("Must be valid PK");
+    ScriptBuf::new_p2tr(&secp, pk, None)
+}
```
nit: as it's common between `bdk_esplora` and `bdk_electrum` too, couldn't it be moved to `bdk_testenv` instead?

evanlinjin replied:
Thanks for pointing this out. I will leave it as it is in the PR to save time.

oleonardolima commented on crates/core/src/spk_client.rs:
```diff
@@ -317,6 +372,13 @@ impl<I> SyncRequest<I> {
+        &mut self,
+    ) -> impl ExactSizeIterator<Item = SpkWithExpectedTxids> + '_ {
+        SyncIter::<I, SpkWithExpectedTxids>::new(self)
+    }
```
question: the `iter_spks` fn above is not being used anymore, at least for `SyncRequest`, do we need to keep it?

evanlinjin replied:
Removed.

oleonardolima reviewed (commented):
cACK 1eb0c322ead9f454093abdfe08bf868f841d4a24

Great work! I left some minor comments/questions that if applicable could be done in a follow-up PR.

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -256,17 +257,30 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+                        ),
+                    )
+                })
+                .unzip();
```
I'm curious about the reasoning behind this

evanlinjin replied:
About what exactly? What the two variables are used for?

ValuedMammal replied:
why split the batch of spks into two new containers?

ValuedMammal replied:
```diff
-            let (spks, expected_spk_txids): (Vec<(u32, ScriptBuf)>, HashMap<ScriptBuf, _>) = spks
-                .iter()
-                .map(|(spk_i, spk_with_expected_txids)| {
-                    (
-                        (*spk_i, spk_with_expected_txids.spk.clone()),
-                        (
-                            spk_with_expected_txids.spk.clone(),
-                            spk_with_expected_txids.expected_txids.clone(),
-                        ),
-                    )
-                })
-                .unzip();
-
             let spk_histories = self
                 .inner
-                .batch_script_get_history(spks.iter().map(|(_, s)| s.as_script()))?;
+                .batch_script_get_history(spks.iter().map(|(_, s)| s.spk.as_script()))?;
 
             for ((spk_index, spk), spk_history) in spks.into_iter().zip(spk_histories) {
                 if spk_history.is_empty() {
                     unused_spk_count = unused_spk_count.saturating_add(1);
                     if unused_spk_count >= stop_gap {
@@ -294,17 +281,15 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
                 let spk_history_set = spk_history
                     .iter()
                     .map(|res| res.tx_hash)
                     .collect::<HashSet<_>>();
 
-                if let Some(expected_txids) = expected_spk_txids.get(&spk) {
-                    tx_update.evicted_ats.extend(
-                        expected_txids
-                            .difference(&spk_history_set)
-                            .map(|&txid| (txid, start_time)),
-                    );
-                }
+                tx_update.evicted_ats.extend(
+                    spk.expected_txids
+                        .difference(&spk_history_set)
+                        .map(|&txid| (txid, start_time)),
+                );
```

</details>

evanlinjin committed:
feat(core)!: Remove redundant `SyncRequest` methods

Methods `next_spk` and `iter_spks` are removed. These are superceded
with `next_spk_with_expected_txids` and `iter_spks_with_expected_txids`.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -16,23 +16,52 @@
+//!     priority than those that are seen earlier. `last_seen` values are transitive. This means
+//!     that the actual `last_seen` value of a transaction is the max of all the `last_seen` values
+//!     of it's descendants.
+//! * `last_evicted` - This is the timestamp of when a transaction is last-seen as evicted in the
```
```suggestion
//! * `last_evicted` - This is the timestamp of when a transaction last went missing from the mempool
```

evanlinjin committed:
feat(examples): Update example crates to use `expected_spk_txids`

evanlinjin commented:
> ut ACK [1eb0c32](https://github.com/bitcoindevkit/bdk/commit/1eb0c322ead9f454093abdfe08bf868f841d4a24)
> 
> Thanks for the fix and all the new docs, looks like a solid solution. Before (or after) this goes in it would be nice to have some confirmation from @ErikDeSmedt that it fixes #1740 for him (at least using electrum and esplora).

@notmandatory 

The tests that we introduce in this PR are based on @ErikDeSmedt's [tests](https://github.com/ErikDeSmedt/bdk-gists/blob/master/tests/wallet.rs) (but we target for esplora/electrum in our PR here).

This PR does not fix evicted-tx-detection for `bdk_bitcoind_rpc`. That work is in a separate PR: #1857.

So I think it is okay to merge this without Erik's confirmation.

LLFourn reviewed (approved):
ACK 0a20724e30316cbb68f7edf3f62028de15d26277

LagginTimes commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -301,6 +319,58 @@ where
+    ///
+    /// This is used to fill [`SyncRequestBuilder::expected_spk_txids`](bdk_core::spk_client::SyncRequestBuilder::expected_spk_txids).
+    ///
+    /// The spk index range can be contrained with `range`.
```
```suggestion
    /// The spk index range can be constrained with `range`.
```

evanlinjin replied:
I wonder how easy it would be to set up CI to check for spelling mistakes. Might save time.

LagginTimes reviewed (commented):
One last nit that I saw.

ValuedMammal commented:
ACK 0a20724e30316cbb68f7edf3f62028de15d26277
